### PR TITLE
Cyrus SASL

### PR DESCRIPTION
**Pull request by aviks**
This does not work on anything other than Linux at this time. Putting it out there to see if someone wants to take this on.

### CODE REVIEW

**Comment by aviks** [comment]
Some products in this case are in subdirectories inside `lib`. eg, ` LibraryProduct("libanonymous", :libanonymous, "lib/sasl2")`

This definition seems to work for Linux, but it does not seem to find the library for the Mac. So I think this is probably a BB problem?

**Comment by giordano** [comment]
> This definition seems to work for Linux, but it does not seem to find the library for the Mac. So I think this is probably a BB problem?

The problem is that the files have the wrong extension:
```
sandbox:${WORKSPACE}/destdir/lib/sasl2 # ls
libanonymous.3.so  libcrammd5.3.so    libdigestmd5.3.so  libntlm.3.so       libotp.3.so        libplain.3.so      libscram.3.so
libanonymous.la    libcrammd5.la      libdigestmd5.la    libntlm.la         libotp.la          libplain.la        libscram.la
libanonymous.so    libcrammd5.so      libdigestmd5.so    libntlm.so         libotp.so          libplain.so        libscram.so
```

**Comment by giordano** [comment]
For Windows, [these patches](https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-cyrus-sasl) might help

**Comment by giordano** [comment]
Still failing with the same error :confused: 
```
[13:14:39] In file included from ../include/saslplug.h:8:0,
[13:14:39]                  from sasldb.h:50,
[13:14:39]                  from allockey.c:52:
[13:14:39] ../include/md5global.h:21:14: error: conflicting types for ‘INT8’
[13:14:39]  typedef long INT8;   /* 64 bits */
[13:14:39]               ^
[13:14:39] In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winnt.h:150:0,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/minwindef.h:163,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windef.h:8,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windows.h:69,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock2.h:23,
[13:14:39]                  from ../config.h:632,
[13:14:39]                  from allockey.c:46:
[13:14:39] /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/basetsd.h:29:23: note: previous declaration of ‘INT8’ was here
[13:14:39]    typedef signed char INT8,*PINT8;
[13:14:39]                        ^
[13:14:39] In file included from ../include/saslplug.h:8:0,
[13:14:39]                  from sasldb.h:50,
[13:14:39]                  from allockey.c:52:
[13:14:39] ../include/md5global.h:25:23: error: conflicting types for ‘UINT8’
[13:14:39]  typedef unsigned long UINT8;  /* 64 bits */
[13:14:39]                        ^
[13:14:39] In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winnt.h:150:0,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/minwindef.h:163,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windef.h:8,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windows.h:69,
[13:14:39]                  from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock2.h:23,
[13:14:39]                  from ../config.h:632,
[13:14:39]                  from allockey.c:46:
[13:14:39] /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/basetsd.h:33:25: note: previous declaration of ‘UINT8’ was here
[13:14:39]    typedef unsigned char UINT8,*PUINT8;
[13:14:39]                          ^
[13:14:39] make[2]: *** [Makefile:515: allockey.lo] Error 1
```
Note this warning during configure:
```
[13:14:39] configure: WARNING: unrecognized options: --without-authdemond
```

**Comment by aviks** [comment]
> Still failing with the same error

I know 😢 

> Note this warning during configure:

that should be fixed now.

File: C/CyrusSASL/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  atomic_patch -p1 ../patches/04-manpage-paths.patch
+  atomic_patch -p1 ../patches/14-MinGW-w64-add-LIBSASL_API-to-function-definitions.patch
+  atomic_patch -p1 ../patches/15-MinGW-w64-define-LIBSASL_EXPORTS_eq_1-for-sasldb.patch
+  atomic_patch -p1 ../patches/16-MinGW-w64-define-WIN32_LEAN_AND_MEAN-avoiding-handle_t-redef.patch
+  atomic_patch -p1 ../patches/17-MinGW-w64-define-S_IRUSR-and-S_IWUSR.patch
+  atomic_patch -p1 ../patches/19-paths-relocation.patch
+  atomic_patch -p1 ../patches/20-mingw-tchar.patch
+  atomic_patch -p1 ../patches/21-fix-getopt-guard.patch
+
+  cp win32/include/md5global.h include/md5global.h
```
Comment (giordano) [inline C/CyrusSASL/build_tarballs.jl:29]: It looks like it's using the original `include/md5global.h` rather than the one that you're copying

Comment (giordano) [reply C/CyrusSASL/build_tarballs.jl:29]: Yeah, this file is deleted during [`make`](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=5878&view=logs&j=e7e877bf-8e27-5cc6-37ef-ebbcef74f0da&t=2de339ae-1031-5827-a669-8653b6053df6&l=1709):
```
[10:28:50] rm -f md5global.h
[10:28:50] ./makemd5 md5global.h
```

**Comment by aviks** [comment]
Hmmm that's _interesting_. The [MINGW-packages build](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-cyrus-sasl/PKGBUILD#L75) does that, so I'm not sure how that works! 😮

**Comment by giordano** [comment]
> Hmmm that's _interesting_. The [MINGW-packages build](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-cyrus-sasl/PKGBUILD#L75) does that, so I'm not sure how that works! 😮 

The `include/Makefile.in` does actually something good (a rare case):
```make
makemd5$(BUILD_EXEEXT) $(makemd5_OBJECTS): CC=$(CC_FOR_BUILD)
makemd5$(BUILD_EXEEXT) $(makemd5_OBJECTS): CFLAGS=$(CFLAGS_FOR_BUILD)
makemd5$(BUILD_EXEEXT): LDFLAGS=$(LDFLAGS_FOR_BUILD)
```
They build the tool `makemdf5` with the host compiler (and the host extension for executable!), which is reasonable since in cross-compilation we can't generally run foreign executables, the problem is that in this way we get the information of the host system, not the target one, and so for Windows it generates a wrong header files.

I don't know how they run this in msys2/MINGW to be honest, they use [the same platform for `host` and `build`](https://github.com/msys2/MINGW-packages/blob/72dac6ecd4dbbe2cbb29ebbb355b6742a9103150/mingw-w64-cyrus-sasl/PKGBUILD#L100-L101).  I take that in their case `makemd.exe` is getting the right information about the target.

**Comment by giordano** [comment]
Congratulations @aviks, this PR has been selected to be the first non-test package to use the new shiny JLL wrappers!

**Comment by aviks** [comment]
I am honoured!

**Pull request merged**